### PR TITLE
Authorization with JWT

### PR DESCRIPTION
**Pull request by Konstantsiy**

### CODE REVIEW

File: internal/auth/jwt.go [hunk 1]
```diff
@@ -1,6 +1,19 @@
...
+import (
+	"fmt"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+)
+
+const (
+	ACTimeout = 12 * time.Hour
+	RTTimeout = 48 * time.Hour
```
Comment (papahmsolo) [inline internal/auth/jwt.go:14]: Can we give this more descriptive names?

File: internal/auth/jwt.go [hunk 2]
```diff
@@ -16,10 +29,42 @@ func NewTokenManager(publicKey, privateKey string) *TokenManager {
...
+	token, err := jwt.Parse(accessToken, func(token *jwt.Token) (i interface{}, err error) {
+		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
+		}
+
+		return []byte(tm.privateKey), nil
+	})
+	if err != nil {
+		return "", err
+	}
```
Comment (papahmsolo) [inline internal/auth/jwt.go:62]: Let's wrap this error giving it some context

File: internal/auth/jwt.go [hunk 3]
```diff
@@ -16,10 +29,42 @@ func NewTokenManager(publicKey, privateKey string) *TokenManager {
...
+
+	claims, ok := token.Claims.(jwt.MapClaims)
```
Comment (papahmsolo) [inline internal/auth/jwt.go:64]: I guess, you can cast it to `jwt.StandardClaims` and return `claims.Subject` or even all claims. Thus, you won't need to case the subject specifically with `claims["sub"].(string)`

File: internal/auth/jwt.go [hunk 4]
```diff
@@ -16,10 +29,42 @@ func NewTokenManager(publicKey, privateKey string) *TokenManager {
...
-	return "", nil
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.StandardClaims{
+		ExpiresAt: time.Now().Add(RTTimeout).Unix(),
+		IssuedAt:  time.Now().Unix(),
+	})
+
+	return token.SignedString(tm.privateKey)
+}
+
+//ParseToken parses the given token and returns the user ID.
```
Comment (papahmsolo) [inline internal/auth/jwt.go:51]: I guess this method is not only parsing the token, but also validating it. Let's add a check for token expiration

File: internal/server/server.go [hunk 5]
```diff
@@ -15,6 +16,15 @@ import (
...
 )
 
+const (
+	AuthorizationHeader  = "Authorization"
+	NeededSecurityScheme = "Bearer"
+)
+
+type key int
+
+const userIDKey key = iota
```
Comment (papahmsolo) [inline internal/server/server.go:26]: Not sure why would we want to use `iota` here, let's just use 0

File: internal/server/server.go [hunk 6]
```diff
@@ -62,8 +72,33 @@ func NewServer(repo *repository.Repository, tokenManager *auth.TokenManager) *Se
...
+		}
+
+		token := authHeaderParts[1]
+		claimsUserID, err := s.tokenManager.ParseToken(token)
+		if err != nil {
+			http.Error(w, "can't parse JWT", http.StatusUnauthorized)
+			return
+		}
+
+		ctx := context.WithValue(r.Context(), userIDKey, claimsUserID)
```
Comment (papahmsolo) [inline internal/server/server.go:99]: Neat! Let's create two functions for this. To put login into context and to retrieve login from it. They'll use `userIDKey` to do it.

Comment (papahmsolo) [reply internal/server/server.go:99]: `func contextWithLogin(ctx context.Context, login string) context.Context`
`func LoginFromContext(ctx context.Context) string`

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good!
Let's make it even better 🙌

File: internal/auth/jwt.go (hunk 1 above)
Comment (Konstantsiy) [reply internal/auth/jwt.go:14]: > Can we give this more descriptive names?

ok, I will describe the constants more detail

File: internal/auth/jwt.go (hunk 3 above)
Comment (Konstantsiy) [reply internal/auth/jwt.go:64]: > I guess, you can cast it to `jwt.StandardClaims` and return `claims.Subject` or even all claims. Thus, you won't need to case the subject specifically with `claims["sub"].(string)`

Thank you for advice!

Comment (Konstantsiy) [reply internal/auth/jwt.go:64]: > I guess, you can cast it to `jwt.StandardClaims` and return `claims.Subject` or even all claims. Thus, you won't need to case the subject specifically with `claims["sub"].(string)`

Thank you for advice!

File: internal/server/server.go (hunk 6 above)
Comment (Konstantsiy) [reply internal/server/server.go:99]: > `func contextWithLogin(ctx context.Context, login string) context.Context`
> `func LoginFromContext(ctx context.Context) string`

OK, but here only user id is used

**Review by papahmsolo — APPROVED** [review]

**Pull request merged**